crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
+            .compute_exchanged_amount(
+                context.client(),
+                context.io(),
+                raw_balance.to_owned(),
+                current_epoch,
+                Conversions::new(),
+            )
+            .await?;
 
-            // re-date the all the latest conversions up one epoch
-            let mut estimated_conversions = Conversions::new();
-            for (asset_type, (conv, wit, _)) in &conversions {
-                let mut asset = match self
-                    .decode_asset_type(context.client(), *asset_type)
-                    .await
+        // Query missing conversions first, i.e. those for assets that carry
+        // the current epoch. There are no conversions for these yet so we
+        // need to see if there are conversions for the previous epoch
+        for (asset_type, _) in raw_balance.components() {
+            let mut asset = match self
+                .decode_asset_type(context.client(), *asset_type)
+                .await
+            {
+                Some(
+                    data @ AssetData {
+                        epoch: Some(ep), ..
+                    },
+                ) if ep == current_epoch => data,
+                _ => continue,
+            };
+            asset.redate(previous_epoch);
+            let redated_asset_type = asset.encode()?;
+
+            self.query_allowed_conversion(
+                context.client(),
+                redated_asset_type,
+                &mut conversions,
+            )
+            .await;
+        }
+
+        let mut estimated_next_epoch_conversions = Conversions::new();
+        // re-date the all the latest conversions up one epoch
+        for (asset_type, (conv, wit, _)) in conversions {
+            let mut asset = match self
+                .decode_asset_type(context.client(), asset_type)
+                .await
+            {
+                Some(
+                    data @ AssetData {
+                        epoch: Some(ep), ..
+                    },
+                ) if ep == previous_epoch => data,
+                _ => continue,
+            };
+            asset.redate_to_next_epoch();
+            let decoded_conv = self
+                .decode_sum(context.client(), conv.clone().into())
+                .await
+                .0;
+            let mut est_conv = I128Sum::zero();
+            for ((_, mut asset_data), val) in decoded_conv.into_components() {
+                // We must upconvert the components of the conversion unless
+                // this conversion is for a non-native token and the component
+                // itself is the native token. In this case we should avoid
+                // upconverting it since conversions for non-native tokens must
+                // always refer to a specific epoch (MaspEpoch(0)) native token
+                if asset.token == native_token
+                    || asset_data.token != native_token
```

Comment:
```suggestion
                if !(asset.token != native_token
                    && asset_data.token == native_token)
```

Reply:
This matches the comment more closely. I was confused until I wrote it out like this.

Reply:
You're right the comment is a bit misleading, I changed that though instead of modifying the if condition